### PR TITLE
coprocessor/dag/expr: make md5 & sha1 a little bit faster.

### PR DESCRIPTION
**Pull request by Hijiao**
Signed-off-by: Max <xiahaijiao@hotmail.com>

## What have you changed? (mandatory)

- using `Md5.result()` and `Sha1.result()` instead of `Md5.result_str()` and `Sha1.result_str()` to malloc an array with explicit capacity on stack instead of the vec with implicit  capacity on heap

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
pass make expression

## Refer to a related PR or issue link (optional)
#3612

### CODE REVIEW

**Comment by solotzg** [comment]
Thanks for your PR. It seems will reduce the unnecessary heap-based memory allocation in `result_str`.

**Comment by DorianZheng** [comment]
@Hijiao Maybe you should benchmark ScalarFunc::md5 and ScalarFunc::sha1, and perform read on the output.

**Comment by Hijiao** [comment]
@DorianZheng I don't think it's necessary to benchmark ScalarFunc::md5 or ScalarFunc::sha1, because my tests have told that the difference between `result()` and `result_str()` is not  noticeable when the capacity is small.  And I am going to test the two different ways in some other encryptions like sha512 which has output of 128 chars.

**Comment by Hijiao** [comment]
@DorianZheng  Hi, I find out the reason why my improvement is a bit faster in theory but not in fact, the function `hex::encode(buff)` which I used  is slower than ` buff[..].to_hex()` used in `result_str()`. After changing it , the `result()` faster 5% than `result_str()` as I expect.

> test tests::md5_using_result_str_test    ... bench:  77,702,181 ns/iter (+/- 5,781,630)
> test tests::md5_using_result_test        ... bench:  73,699,068 ns/iter (+/- 4,354,432)
> test tests::sha1_using_result_str_test   ... bench:  76,174,790 ns/iter (+/- 5,049,940)
> test tests::sha1_using_result_test       ... bench:  70,833,823 ns/iter (+/- 4,416,995)
> test tests::sha512_using_result_str_test ... bench: 116,338,958 ns/iter (+/- 5,704,834)
> test tests::sha512_using_result_test     ... bench: 113,144,491 ns/iter (+/- 4,888,486)
> 

```rust
#![feature(test)]
extern crate crypto;
extern crate test;
extern crate hex;
extern crate rustc_serialize;

use rustc_serialize::hex::ToHex;
use crypto::{digest::Digest, md5::Md5, sha1::Sha1, sha2::Sha512};

fn md5_using_result_str(input: &[u8]) -> Vec<u8> {
    let mut hasher = Md5::new();
    hasher.input(input);
    hasher.result_str().into_bytes()
}

fn md5_using_result(input: &[u8]) -> Vec<u8> {
    let mut hasher = Md5::new();
    let mut buff: [u8; 16] = [0; 16];
    hasher.input(input);
    hasher.result(&mut buff);
    buff[..].to_hex().into_bytes()
}

fn sha1_using_result_str(input: &[u8]) -> Vec<u8> {
    let mut hasher = Sha1::new();
    hasher.input(input.as_ref());
    hasher.result_str().into_bytes()
}

fn sha1_using_result(input: &[u8]) -> Vec<u8> {
    let mut hasher = Sha1::new();
    hasher.input(input.as_ref());
    let mut buff: [u8; 20] = [0; 20];
    hasher.result(&mut buff);
    buff[..].to_hex().into_bytes()
}

fn sha512_using_result_str(input: &[u8]) -> Vec<u8> {
    let mut hasher = Sha512::new();
    hasher.input(input.as_ref());
    hasher.result_str().into_bytes()
}

fn sha512_using_result(input: &[u8]) -> Vec<u8> {
    let mut hasher = Sha512::new();
    hasher.input(input.as_ref());
    let mut buff: [u8; 64] = [0; 64];
    hasher.result(&mut buff);
    buff[..].to_hex().into_bytes()
}

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;
    
    const TOTAL_TEST: i32 = 100_000;

    #[bench]
    fn md5_using_result_str_test(b: &mut Bencher) {
        b.iter(|| {
            for i in 0..TOTAL_TEST {
                let s = format!("{:032}", i);
                md5_using_result_str(s.as_ref());
            }
        });
    }

    #[bench]
    fn md5_using_result_test(b: &mut Bencher) {
        b.iter(|| {
            for i in 0..TOTAL_TEST {
                let s = format!("{:032}", i);
                md5_using_result(s.as_ref());
            }
        });
    }


    #[bench]
    fn sha1_using_result_str_test(b: &mut Bencher) {
        b.iter(|| {
            for i in 0..TOTAL_TEST {
                let s = format!("{:032}", i);
                sha1_using_result_str(s.as_ref());
            }
        });
    }

    #[bench]
    fn sha1_using_result_test(b: &mut Bencher) {
        b.iter(|| {
            for i in 0..TOTAL_TEST {
                let s = format!("{:032}", i);
                sha1_using_result(s.as_ref());
            }
        });
    }


    #[bench]
    fn sha512_using_result_str_test(b: &mut Bencher) {
        b.iter(|| {
            for i in 0..TOTAL_TEST {
                let s = format!("{:032}", i);
                sha512_using_result_str(s.as_ref());
            }
        });
    }

    #[bench]
    fn sha512_using_result_test(b: &mut Bencher) {
        b.iter(|| {
            for i in 0..TOTAL_TEST {
                let s = format!("{:032}", i);
                sha512_using_result(s.as_ref());
            }
        });
    }
}
```

File: src/coprocessor/dag/expr/builtin_encryption.rs [hunk 1]
```diff
@@ -25,8 +25,10 @@ impl ScalarFunc {
     ) -> Result<Option<Cow<'a, [u8]>>> {
         let input = try_opt!(self.children[0].eval_string(ctx, row));
         let mut hasher = Md5::new();
+        let mut buff: [u8; 16] = [0; 16];
         hasher.input(input.as_ref());
-        let md5 = hasher.result_str().into_bytes();
+        hasher.result(&mut buff);
+        let md5 = buff[..].to_hex().into_bytes();
```
Comment (hicqu) [inline src/coprocessor/dag/expr/builtin_encryption.rs]: So there is still a copy. I think alloc 16 bytes directly on heap is better.

File: src/lib.rs [hunk 2]
```diff
@@ -73,6 +73,7 @@ extern crate raft;
 extern crate rand;
 extern crate regex;
 extern crate rocksdb;
+extern crate rustc_serialize;
```
Comment (BusyJay) [inline src/lib.rs]: Better not use an internal crate.

File: src/coprocessor/dag/expr/builtin_encryption.rs (hunk 1 above)
Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_encryption.rs]: I have checked `to_hex` implement. It's ok. Sorry.

File: src/lib.rs (hunk 2 above)
Comment (Hijiao) [reply src/lib.rs]: rustc_serialize is not an internal crate,  see https://github.com/tikv/tikv/blob/d642e43273850100ea6d37e2c292d164a0b0f355/Cargo.toml#L70

**Comment by solotzg** [comment]
@Hijiao please help to resolve conflicts and format. Thanks.

**Comment by solotzg** [comment]
@breeswish PTAL

**Pull request merged**